velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -60,20 +61,148 @@ void WindowFuzzer::addWindowFunctionSignatures(
+      core::WindowNode::BoundType::kCurrentRow,
+      core::WindowNode::BoundType::kFollowing};
+  endBoundOptions = {
+      core::WindowNode::BoundType::kPreceding,
+      core::WindowNode::BoundType::kCurrentRow,
+      core::WindowNode::BoundType::kFollowing,
+      core::WindowNode::BoundType::kUnboundedFollowing};
+
+  // End bound option should not be greater than start bound option as this
+  // would result in an invalid frame.
+  auto startBoundIndex = boost::random::uniform_int_distribution<uint32_t>(
+      0, startBoundOptions.size() - 1)(rng_);
+  auto endBoundMinIdx = std::max(0, static_cast<int>(startBoundIndex) - 1);
+  auto endBoundIndex = boost::random::uniform_int_distribution<uint32_t>(
+      endBoundMinIdx, endBoundOptions.size() - 1)(rng_);
+  auto frameStartBoundType = startBoundOptions[startBoundIndex];
+  auto frameEndBoundType = endBoundOptions[endBoundIndex];
+
+  return std::make_tuple(windowType, frameStartBoundType, frameEndBoundType);
+}
+
+// For frames with k RANGE PRECEDING/FOLLOWING, Velox requires the application
+// to add columns with the range frame boundary value computed according to the
+// frame type.
+// If the frame is k PRECEDING :
+// frame_boundary_value = current_order_by - k (for ascending ORDER BY)
+// frame_boundary_value = current_order_by + k (for descending ORDER BY)
+// If the frame is k FOLLOWING :
+// frame_boundary_value = current_order_by + k (for ascending ORDER BY)
+// frame_boundary_value = current_order_by - k (for descending ORDER BY)
+template <typename T>
+T WindowFuzzer::genOffsetAtIdx(
+    const T& offsetColumnVal,
+    T offsetValue,
+    core::WindowNode::BoundType frameBoundType,
+    core::SortOrder sortOrder) {
+  if ((frameBoundType == core::WindowNode::BoundType::kPreceding &&
+       sortOrder.isAscending()) ||
+      (frameBoundType == core::WindowNode::BoundType::kFollowing &&
+       !sortOrder.isAscending())) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      return offsetColumnVal - offsetValue;
+    } else {
+      return checkedMinus<T>(offsetColumnVal, offsetValue);
     }
-  };
-  auto isRowsFrame =
-      boost::random::uniform_int_distribution<uint32_t>(0, 1)(rng_);
-  auto frameTypeString = frameType(isRowsFrame);
+  }
+
+  if ((frameBoundType == core::WindowNode::BoundType::kFollowing &&
+       sortOrder.isAscending()) ||
+      (frameBoundType == core::WindowNode::BoundType::kPreceding &&
+       !sortOrder.isAscending())) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      return offsetColumnVal + offsetValue;
+    } else {
+      return checkedPlus<T>(offsetColumnVal, offsetValue);
+    }
+  }
+
+  VELOX_UNREACHABLE(
+      "Offset cannot be generated: sortOrder ascending {}, frameBoundType {}",
+      sortOrder.toString(),
+      core::WindowNode::boundTypeName(frameBoundType));
+}
+
+template <typename T>
+std::optional<std::string> WindowFuzzer::addKRangeFrameColumnToInput(
```

Comment:
We should ensure that we send a column of type that can be handled in this function. Returning nullOpt is weird. We are generating columns of the correct types. Just adding a VELOX_CHECK for the type here is sufficient.